### PR TITLE
Feat/54 add an introduction textarea

### PR DESCRIPTION
**Pull request by grnd-alt**
* Resolves: #54 
* Target version: main

### Summary
introduce a description editor at the top similar to Files
#### Screenshots:

Viewing a Table:
image

Creating a Table:
image

Editing a Table:
image

### Checklist

- [ ] Code is properly formatted
- [ ] Sign-off message is added to all commits
- [ ] Tests (unit, integration, api and/or acceptance) are included
- [ ] Documentation (manuals or wiki) has been updated or is not required

### CODE REVIEW

**Comment by juliushaertl** [comment]
Very nice, two thoughts that might be worth to discuss with @nextcloud/designers :
- The editor is not collaborative, so maybe we want to make it read only by default with a button to toggle to edit
- From my perspective it would be nicer to left-align as we do it in the files app with the folder description

**Review by jancborchardt — COMMENTED** [review]
> - The editor is not collaborative, so maybe we want to make it read only by default with a button to toggle to edit
> - From my perspective it would be nicer to left-align as we do it in the files app with the folder description

Agree with both of @juliushaertl’s points. Actually a great example of how this could look is the Forms app:

## View mode
image

## Edit mode
image

**Review by jancborchardt — CHANGES_REQUESTED** [review]
@grnd-alt nice, thanks for both the suggestions! :)

- A little bit of moving the description can already achieve a lot, like below when the text is left-aligned with the text of the table below (see modified screenshot below)
- The radio button is a bit huge there and not super needed. I would say an "Edit"/"Done" button in the top right like we have in Collectives is better in this case.
- Is "Hide" a requirement? If so, we can do that via a simple collapsing arrow like we do in Collectives too for the "Recent pages" section
- Can we somehow hide the "Show outline" button in view mode, it is not needed here. cc @juliushaertl
 
316519952-a2efb4bc-ee82-44fd-aa80-f4408ccef8ed

File: src/modules/main/sections/TableDescription.vue [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+						return
+					}
+					this.description = markdown
+					this.updateDescription()
+				},
+			})
+			this.editor.setReadOnly(true)
+		},
+		async saveDescription() {
+			if (this.descriptionLastEdited !== 0 || this.description === this.activeElement.description || !this.canManageElement(this.activeElement)) return
```
Comment (juliushaertl) [inline src/modules/main/sections/TableDescription.vue:94]: Generally preferred code style around the Nextcloud code base and in this case also more readable I think ;)
```suggestion
			if (this.descriptionLastEdited !== 0 || this.description === this.activeElement.description || !this.canManageElement(this.activeElement)) {
				return
			}
```

File: src/modules/main/sections/TableDescription.vue [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+			this.descriptionSaving = true
+			await this.$store.dispatch('updateTableProperty', { id: this.activeElement.id, data: { description: this.description }, property: 'description' })
+			this.descriptionLastEdit = 0
+			this.descriptionSaving = false
+		},
+		updateDescription() {
+			this.descriptionLastEdit = Date.now()
+			clearTimeout(this.descriptionSaveTimeout)
+			this.descriptionSaveTimeout = setTimeout(async () => {
```
Comment (juliushaertl) [inline src/modules/main/sections/TableDescription.vue:103]: Out of curiosity: is setTimeout actually handling async functions properly?

**Review by juliushaertl — CHANGES_REQUESTED** [review]
Small comments, otherwise this looks good 👍

File: src/modules/main/sections/TableDescription.vue (hunk 2 above)
Comment (grnd-alt) [reply src/modules/main/sections/TableDescription.vue]: I don't think setTimeout actually cares what type of function is passed. After the timer has expired it just calls it and has no influence on it anyways. Though wrapping the actual function call in a async arrow function was useless in this case, I removed that.

**Comment by enjeck** [comment]
@jancborchardt I find it very confusing to have an edit button just for the single "description" field. Then a separate button that edits the title in a separate place. Then no way to add a description when creating a table. As a user, I would expect to be able to add a description when creating a table. And to either have a single button to edit everything at once or to be able to inline edit everything. Having a mix feels weird. Maybe it's just me

**Comment by jancborchardt** [comment]
> I find it very confusing to have an edit button just for the single "description" field. Then a separate button that edits the title in a separate place. Then no way to add a description when creating a table. As a user, I would expect to be able to add a description when creating a table. And to either have a single button to edit everything at once or to be able to inline edit everything. Having a mix feels weird. Maybe it's just me

@enjeck yeah you are right – so then should we move the description box editing (and adding) into the modal? And on the tables page we show it read-only, just like we do for the title?

File: lib/Migration/Version000900Date20240314000000.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+	 * @param Closure $schemaClosure The `\Closure` returns a `ISchemaWrapper`
+	 * @param array $options
+	 * @return null|ISchemaWrapper
+	 * @throws Exception
+	 */
+	public function changeSchema(IOutput $output, Closure $schemaClosure, array $options): ?ISchemaWrapper {
+		/** @var ISchemaWrapper $schema */
+		$schema = $schemaClosure();
+
+		if ($schema->hasTable('tables_tables')) {
```
Comment (blizzz) [inline lib/Migration/Version000900Date20240314000000.php:28]: It should also check whether it has the column already, and only add it if not present.

Also, in the migration file where this table is added, the column has to be introduced, too.

**Review by blizzz — CHANGES_REQUESTED** [review]

File: src/store/store.js [hunk 4]
```diff
@@ -262,11 +262,27 @@ export default new Vuex.Store({
 			}
 			return true
 		},
+		async updateTableProperty({ state, commit, dispatch }, { id, data, property }) {
```
Comment (enjeck) [inline src/store/store.js:265]: I can't find where this function is being used? Maybe it is no longer needed?

Comment (grnd-alt) [reply src/store/store.js]: You're right that's an artifact from when the description was editable in the tables view.

**Comment by enjeck** [comment]
This looks good to me too. Thanks, @grnd-alt

**Comment by blizzz** [comment]
@juliushaertl your change request is still active

**Review by juliushaertl — APPROVED** [review]

**Pull request merged**